src/aind_data_asset_indexer/codeocean_bucket_indexer.py

```diff
@@ -44,6 +47,169 @@ def __init__(self, job_settings: CodeOceanIndexBucketJobSettings):
+        -------
+        List[str]
+
+        """
+        external_links = docdb_record.get("external_links", [])
+
+        # Hopefully, ExternalPlatforms.CODEOCEAN doesn't change
+        if isinstance(external_links, dict):
+            external_links = external_links.get(
+                ExternalPlatforms.CODEOCEAN.value, []
+            )
+        else:
+            external_links = [
+                r.get(ExternalPlatforms.CODEOCEAN.value)
+                for r in external_links
+            ]
+        return external_links
+
+    def _update_external_links_in_docdb(
+        self, docdb_client: MongoClient
+    ) -> None:
+        """
+        This method will:
+        1) Retrieve a list of codeocean data asset ids and locations from CO
+        2) Paginate through the docdb records where the location doesn't match
+        the internal co bucket.
+        3) Add or remove the external_links from the docdb record if needed.
+        Parameters
+        ----------
+        docdb_client : MongoClient
+
+        Returns
+        -------
+        None
+
+        """
+        # Should return a list like [{"id": co_id, "location": "s3://..."},]
+        list_of_co_ids_and_locations = self._get_external_data_asset_records()
+        db = docdb_client[self.job_settings.doc_db_db_name]
+        collection = db[self.job_settings.doc_db_collection_name]
+        if list_of_co_ids_and_locations is not None:
+            co_loc_to_id_map = self._map_external_list_to_dict(
+                list_of_co_ids_and_locations
+            )
+            pages = paginate_docdb(
+                docdb_client=docdb_client,
+                db_name=self.job_settings.doc_db_db_name,
+                collection_name=self.job_settings.doc_db_collection_name,
+                filter_query={
+                    "location": {
+                        "$not": {
+                            "$regex": f"^s3://{self.job_settings.s3_bucket}.*"
+                        }
+                    }
+                },
+                projection={"_id": 1, "location": 1, "external_links": 1},
+                page_size=500,
+            )
+            for page in pages:
+                records_to_update = []
+                for record in page:
+                    location = record.get("location")
+                    external_links = self._get_co_links_from_record(record)
+                    code_ocean_ids = (
+                        None
+                        if location is None
+                        else co_loc_to_id_map.get(location)
+                    )
+                    docdb_rec_id = record["_id"]
+                    if (
+                        external_links is not None
+                        and code_ocean_ids is not None
+                        and code_ocean_ids != set(external_links)
+                    ):
+                        new_external_links = code_ocean_ids
+                    elif external_links is not None and not code_ocean_ids:
```

Comment:
Is this part to remove/reset the `external_links` if code ocean api did not return any for that location? Maybe we should log something here in case there are issues with many links being removed.

Reply:
added log info line